### PR TITLE
Align credential schema attribute handling in provisioning

### LINKED ISSUE #2633: Persist schema-defined credential attributes during registration provisioning
### Current Limitation

During registration flows, credential attributes can be prompted from the user based on the schema, but provisioning may still miss them when creating the user. This happens because credential persistence is limited to node inputs instead of using the full schema-defined credential attribute set.

### Suggested Improvement

Make credential provisioning schema-backed as well. Use credential attribute metadata from the user schema, with a requiredOnly option similar to non-credential attributes, so registration flows consistently prompt and persist credential fields defined in the schema.

### PR DESCRIPTION
**Pull request by Dilusha-Madushan**
### Purpose
This PR fixes registration provisioning when credential attributes are defined dynamically in the user schema. Previously, credential inputs could be prompted to the user but then be missed during user provisioning because the executor only persisted credential attributes based on node inputs instead of the schema-backed credential set.

This change also makes credential schema handling consistent with non-credential schema handling by exposing credential attribute metadata with a `requiredOnly` option and using that metadata in the provisioning executor when building dynamic prompt inputs.

### Approach
Updated the userschema contract to support `GetCredentialAttributeInfos(..., requiredOnly)` similar to `GetNonCredentialAttributes(..., requiredOnly)`, and propagated that change through the service layer and generated mocks.

In the provisioning executor, credential prompting and credential persistence are now handled as two explicit paths:
- Prompt-time credential resolution uses credential schema metadata, including required/display name information, so dynamic prompt inputs are built correctly.
- Provisioning-time credential persistence fetches schema credential attribute names and persists any available values from the execution context instead of limiting persistence to credential node inputs only.

The executor logic was also clarified by separating prompt-side non-credential and credential schema fetch helpers, and tests were expanded to cover:
- optional credential overridden to required by node input
- capped forwarded dynamic prompt inputs
- credential attribute persistence during provisioning
- required-only vs all-attributes behavior for credential schema metadata in schema and service layers

### Related Issues
- https://github.com/asgardeo/thunder/issues/2633

### Related PRs
- https://github.com/asgardeo/thunder/pull/2537

## Summary by CodeRabbit

* **New Features**
  * Enhanced provisioning prompts: separate required/optional credential and non-credential pools with max-per-prompt batching.
  * Improved password-field synchronization for forwarded inputs.

* **Bug Fixes**
  * Prevent re-prompting of already-present optional attributes.
  * Ensure promotions to password type don't mutate shared meta across runs.
  * Better error handling when credential attribute retrieval fails.

* **Tests**
  * Expanded tests for credential handling, prompting limits, and edge cases.

### CODE REVIEW

File: backend/internal/flow/core/prompt_node.go [hunk 1]
```diff
@@ -326,6 +326,12 @@ func (n *promptNode) enrichInputsFromForwardedData(ctx *NodeContext, nodeResp *c
 				n.logger.Debug("Updated input required flag from ForwardedData",
 					log.String("identifier", fwdInput.Identifier))
 			}
+			if fwdInput.Type == common.InputTypePassword &&
+				nodeResp.Inputs[idx].Type != common.InputTypePassword {
+				nodeResp.Inputs[idx].Type = common.InputTypePassword
+				n.logger.Debug("Updated input type to password from ForwardedData",
+					log.String("identifier", fwdInput.Identifier))
+			}
```
[the remark this answers is not in the document]
Comment (Dilusha-Madushan) [reply backend/internal/flow/core/prompt_node.go:456]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go [hunk 2]
```diff
@@ -279,65 +285,121 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
+			continue
+		}
+		nodeInp, inNodeInputs := nodeInputMap[attr.Attribute]
+		effectiveRequired := attr.Required
+		if inNodeInputs {
+			effectiveRequired = attr.Required || nodeInp.Required
+		}
+		if !effectiveRequired && !inNodeInputs {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go [hunk 3]
```diff
@@ -279,65 +285,121 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
 	}
 
-	if maxInputs := p.getMaxDynamicInputs(ctx); maxInputs > 0 && len(schemaMissingAttrs) > maxInputs {
-		schemaMissingAttrs = schemaMissingAttrs[:maxInputs]
+	// Apply maxPerPrompt cap only to ForwardedData — controls how many attrs are shown per prompt
+	// round. Required fills first so the cap trims from the optional end.
+	toForward := allSchemaMissing
+	if maxInputs := p.getMaxDynamicInputs(ctx); maxInputs > 0 && len(toForward) > maxInputs {
+		toForward = toForward[:maxInputs]
 	}
```
[the remark this answers is not in the document]
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go [hunk 4]
```diff
@@ -279,65 +284,123 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
+	// iterations know not to re-prompt them if the user left them empty.
+	p.storePresentedOptionalAttrs(execResp, toForward, alreadyPromptedOptionalAttrs)
 
-	execResp.Inputs = upsertInputs(execResp.Inputs, schemaMissingAttrs)
+	execResp.Inputs = allSchemaMissing
 	if execResp.ForwardedData == nil {
 		execResp.ForwardedData = make(map[string]interface{})
 	}
-	execResp.ForwardedData[common.ForwardedDataKeyInputs] = schemaMissingAttrs
+	execResp.ForwardedData[common.ForwardedDataKeyInputs] = toForward
```
[the remark this answers is not in the document]
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go:343]: @coderabbitai 

The suggestion is not applicable to this implementation.

The provisioning executor is schema-driven — it can only collect and store attributes defined by the entity type. nodeInputMap is not used to introduce independent inputs; it exists solely to apply the optional → required upgrade rule on schema attributes (effectiveRequired = attr.Required || nodeInp.Required) and to gate which optional credential attributes are prompted. Node-defined inputs that have no corresponding entry in schemaAttrs or credAttrs are intentionally ignored because there is no schema field to store their value against.

checkNodeInputs has since been removed for this reason. The execResp.Inputs = allSchemaMissing assignment is correct: allSchemaMissing already reflects any required overrides from node inputs via effectiveRequired, so nothing meaningful is dropped. Merging with upsertInputs would surface schema-absent node inputs in the prompt, which the provisioning executor can never store.

File: backend/internal/flow/executor/provisioning_executor.go [hunk 5]
```diff
@@ -276,101 +281,155 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
+		}
+		if !effectiveRequired && alreadyPromptedOptionalAttrs[attr.Attribute] {
 			continue
 		}
 		input := common.Input{
 			Identifier:  attr.Attribute,
 			Type:        common.InputTypeText,
-			Required:    attr.Required,
+			Required:    effectiveRequired,
 			DisplayName: attr.DisplayName,
```
[the remark this answers is not in the document]
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/entitytype/service.go [hunk 6]
```diff
@@ -77,6 +77,9 @@ type EntityTypeServiceInterface interface {
 	GetCredentialAttributes(
 		ctx context.Context, category TypeCategory, entityType string,
 	) ([]string, *serviceerror.ServiceError)
+	GetCredentialAttributeInfos(
```
Comment (ThaminduDilshan) [inline backend/internal/entitytype/service.go]: Rather than introducing a new `GetCredentialAttributeInfos()` method, shall we improve the existing `GetCredentialAttributes()` to return an array of AttributeInfo?

File: backend/internal/entitytype/model/schema.go [hunk 7]
```diff
@@ -142,6 +142,27 @@ func (cs *Schema) GetNonCredentialAttributes(requiredOnly bool) []AttributeInfo
 	return result
 }
 
+// GetCredentialAttributeInfos returns top-level properties marked as credentials with their
+// required status and display names. When requiredOnly is true, only required properties are
+// included.
+func (cs *Schema) GetCredentialAttributeInfos(requiredOnly bool) []AttributeInfo {
```
Comment (ThaminduDilshan) [inline backend/internal/entitytype/model/schema.go]: Follow the same approach proposed in previous comment. Let's update existing `GetCredentialAttributes()` method

File: backend/internal/flow/executor/provisioning_executor.go [hunk 8]
```diff
@@ -264,9 +264,14 @@ func (p *provisioningExecutor) handleExistingUser(ctx *core.NodeContext, userID
...
-// missing inputs to the executor response. Returns true if all required inputs — both
-// node-defined and schema-derived — are satisfied, otherwise false.
+// HasRequiredInputs checks whether all schema-driven provisioning inputs are satisfied and appends
+// any missing promptable schema attrs to the executor response. Node inputs influence requiredness
+// and prompt metadata for schema attrs, but schema-absent node inputs are ignored.
+//
+// Missing inputs are ordered as: required credentials -> required non-credentials ->
+// optional credentials -> optional non-credentials. maxPerPrompt caps the forwarded
+// prompt batch after this list is built. includeOptional only affects optional
+// non-credential attrs.
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/provisioning_executor.go:278]: Shouldn't we order credentials to come after non credential attributes?

File: backend/internal/flow/executor/provisioning_executor.go [hunk 9]
```diff
@@ -276,101 +281,163 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
-	schemaAttrs, err := p.fetchSchemaAttributes(ctx, logger)
+	// Fetch non-credential and credential attributes from the user type.
+	schemaAttrs, err := p.fetchPromptNonCredentialSchemaAttrs(ctx, logger)
 	if err != nil {
 		execResp.Status = common.ExecFailure
 		return false
 	}
-	if len(schemaAttrs) == 0 {
-		return nodeInputsSatisfied
+	credAttrs, err := p.fetchPromptCredentialSchemaAttrs(ctx, logger)
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/provisioning_executor.go]: Rather than fetching the schema twice for credential and non-credential attributes, shall we improve this to fetch them at once?

Thinking whether we need to introduce another new method to the schema service.
1. We can fetch the entire schema and process it here rather than introducing a new method.
2. Introduce a new method to the schema service to process and return attributes as `[]AttributeInfo`.
    - Maybe we can introduce a boolean credential field to the `AttributeInfo` struct and take credential attributes filtering as a method parameter
    - GetAttributes(ctx context.Context, category TypeCategory, entityType string, requiredOnly bool, credentialsOnly bool) ([]AttributeInfo, *serviceerror.ServiceError)

In a way it's better to handover this processing responsibility to the schema service. Hence I prefer Option 2

File: backend/internal/entitytype/service.go (hunk 6 above)
Comment (ThaminduDilshan) [reply backend/internal/entitytype/service.go]: Since you're anyway going to change the existing method, we can replace the usage of this `GetCredentialAttributes()` method with the new method that we're introducing here. Just have to pass credentials only flag to true

https://github.com/asgardeo/thunder/pull/2553#discussion_r3206050838

File: backend/internal/flow/executor/provisioning_executor.go (hunk 9 above)
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/provisioning_executor.go]: So logic of `HasRequiredInputs()` method would be;
1. Fetch the entire schema
2. Get already prompted optional attributes from runtime
3. Rather than doing two passes in `buildNonCredentialMissingInputs` and `buildCredentialMissingInputs`, find the ncRequiredMissing, ncOptionalMissing, credRequiredMissing and credOptionalMissing attributes in a single pass. No need to have two such methods
4. Build full schema missing list. Only append ncOptionalMissing and credOptionalMissing if prompt optional enabled
5. Apply max per prompt filtering
6. Record prompted optional attributes

WDYT? This optimizes the current provisioning executor implementation as well

File: backend/internal/flow/executor/provisioning_executor.go [hunk 10]
```diff
@@ -276,101 +281,163 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
+		execResp.ForwardedData = make(map[string]interface{})
+	}
+	execResp.ForwardedData[common.ForwardedDataKeyInputs] = toForward
+	logger.Debug("Schema attributes are missing, requesting via prompt",
+		log.Int("missingCount", len(allSchemaMissing)))
+	return false
+}
 
-	// Build required and optional missing lists separately so required ones are always shown first.
+func (p *provisioningExecutor) buildNonCredentialMissingInputs(
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/provisioning_executor.go]: Add missing method comments

File: backend/internal/flow/executor/provisioning_executor.go [hunk 11]
```diff
@@ -543,24 +631,7 @@ func (p *provisioningExecutor) appendCredentialAttributes(ctx *core.NodeContext,
...
-
-	attrsToPopulate := schemaCredAttrs
-	if len(nodeCredentialInputs) > 0 {
-		attrsToPopulate = nodeCredentialInputs
-	}
-
-	for _, attr := range attrsToPopulate {
 		if value, exists := ctx.UserInputs[attr]; exists {
 			(*attributes)[attr] = value
 		} else if runtimeValue, exists := ctx.RuntimeData[attr]; exists {
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/provisioning_executor.go]: Aligning with my previous comment, we can improve the `getAttributesForProvisioning()` method to return two maps; identifying and credential attributes separately. Then the caller can try to identify user with the identifying attributes and append two maps at the end.
This will remove another iteration over the schema attributes

File: backend/internal/flow/core/prompt_node.go [hunk 12]
```diff
@@ -529,44 +535,44 @@ func filterMetaComponents(comps []interface{}, allowedRefs, knownInputActionRefs
...
-
-	// Build a set of refs/ids from the node's own configured prompt inputs —
-	// used to suppress synthesis for node-defined inputs with no meta component.
-	nodeInputRefs := make(map[string]struct{})
-	for _, inp := range n.getAllInputs() {
-		if inp.Ref != "" {
-			nodeInputRefs[inp.Ref] = struct{}{}
-		}
-		nodeInputRefs[inp.Identifier] = struct{}{}
+func cloneBlockWithChildren(compMap map[string]interface{}, children []interface{}) map[string]interface{} {
```
Comment (ThaminduDilshan) [inline backend/internal/flow/core/prompt_node.go:663]: Append missing method comments. Check other places too

File: backend/internal/entitytype/service.go (hunk 6 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/service.go]: @coderabbitai check

File: backend/internal/entitytype/model/schema.go (hunk 7 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/model/schema.go]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go (hunk 8 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go:278]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go (hunk 9 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go (hunk 10 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/flow/executor/provisioning_executor.go (hunk 11 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check

File: backend/internal/flow/core/prompt_node.go (hunk 12 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/core/prompt_node.go:663]: @coderabbitai check

File: backend/internal/entitytype/service.go (hunk 6 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/service.go]: @coderabbitai I addressed the refactoring, check the current status

File: backend/internal/entitytype/model/schema.go (hunk 7 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/model/schema.go]: @coderabbitai check is this still exist or already fixed?

File: backend/internal/entitytype/service.go [hunk 13]
```diff
@@ -74,18 +74,16 @@ type EntityTypeServiceInterface interface {
 		attributes json.RawMessage,
 		exists func(map[string]interface{}) (bool, error),
 	) (bool, *serviceerror.ServiceError)
-	GetCredentialAttributes(
+	GetAttributeInfos(
```
Comment (ThaminduDilshan) [inline backend/internal/entitytype/service.go]:
```suggestion
	GetAttributes(
```

File: backend/internal/entitytype/model/schema.go [hunk 14]
```diff
@@ -113,21 +113,27 @@ func (cs *Schema) ValidateAsDisplayAttribute(name string) DisplayAttributeStatus
...
+	Credential  bool
 }
 
-// GetNonCredentialAttributes returns top-level properties not marked as credentials.
-// When requiredOnly is true, only required properties are included.
-func (cs *Schema) GetNonCredentialAttributes(requiredOnly bool) []AttributeInfo {
+// GetAttributeInfos returns top-level properties filtered by the provided flags.
+// allowCredential includes credential properties; allowNonCredential includes non-credential
+// properties. When requiredOnly is true, only required properties are included.
+func (cs *Schema) GetAttributeInfos(allowCredential, allowNonCredential, requiredOnly bool) []AttributeInfo {
```
Comment (ThaminduDilshan) [inline backend/internal/entitytype/model/schema.go]:
```suggestion
func (cs *Schema) GetAttributes(allowCredential, allowNonCredential, requiredOnly bool) []AttributeInfo {
```

File: backend/internal/flow/executor/provisioning_executor.go [hunk 15]
```diff
@@ -276,102 +285,142 @@ func (p *provisioningExecutor) HasRequiredInputs(ctx *core.NodeContext,
...
 }
 
-// fetchSchemaAttributes retrieves non-credential attributes from the entity type service.
-// When promptOptionalAttributes is true it fetches all attributes; otherwise only required ones.
-func (p *provisioningExecutor) fetchSchemaAttributes(
-	ctx *core.NodeContext, logger *log.Logger,
+// fetchSchemaAttributeInfos retrieves schema attributes from the entity type service for the
+// current user type. allowCredential and allowNonCredential control which attribute classes
+// are returned.
+func (p *provisioningExecutor) fetchSchemaAttributeInfos(
```
Comment (ThaminduDilshan) [inline backend/internal/flow/executor/provisioning_executor.go]:
```suggestion
func (p *provisioningExecutor) fetchSchemaAttributes(
```

File: backend/internal/entitytype/service.go (hunk 13 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/service.go]: @coderabbitai check changes

File: backend/internal/entitytype/model/schema.go (hunk 14 above)
Comment (Dilusha-Madushan) [reply backend/internal/entitytype/model/schema.go]: @coderabbitai check changes

File: backend/internal/flow/executor/provisioning_executor.go (hunk 15 above)
Comment (Dilusha-Madushan) [reply backend/internal/flow/executor/provisioning_executor.go]: @coderabbitai check changes

**Review by ThaminduDilshan — APPROVED** [review]
